src/compiler.rkt

```diff
@@ -219,3 +194,168 @@
 ;; Like `compile-progs`, but a single prog.
 (define (compile-prog expr ctx)
   (compose first (compile-progs (list expr) ctx)))
+
+(define (backward-pass ivec varc vregs vprecs vstart-precs)
+  (vector-fill! vprecs 0)
+  (for ([instr (in-vector ivec (- (vector-length ivec) 1) -1 -1)]
+        [n (in-range (- (vector-length vregs) 1) -1 -1)])
+
+    (define op (car instr))
+    (define tail-registers (rest instr))
+    (define srcs (map (lambda (x) (vector-ref vregs x)) tail-registers))
+    (define output (vector-ref vregs n))
+
+    (define exps-from-above (vector-ref vprecs n))
+    (define final-parent-precision (min (*max-mpfr-prec*) (+ exps-from-above (vector-ref vstart-precs n))))
+    (when (equal? final-parent-precision (*max-mpfr-prec*))
+      (*sampling-iteration* (*max-sampling-iterations*)))
+    (vector-set! vprecs n final-parent-precision)
+
+    (define new-exponents 0)
+    (cond
+      [(equal? op ival-add)
+       (define xlo (ival-lo (first srcs)))
+       (define xlo-exp (true-exponent xlo))
+       (define xlo-sgn (bigfloat-signbit xlo))
+       (define ylo (ival-lo (second srcs)))
+       (define ylo-exp (true-exponent ylo))
+       (define ylo-sgn (bigfloat-signbit ylo))
+       (define outlo-exp (true-exponent (ival-lo output)))
+       ; consider only lower bound
+       (set! new-exponents
```

Comment:
In principle this would be a little more accurate if it returned separate new exponents for each thing being added, right? Each one based on its own condition number?

Reply:
Yes, that's would be more correct

Reply:
Ok. I'd like to try it, but it can be a separate PR. Basically I think it's possible it'll speed things up a little bit.